lib/3scale/backend/stats/cleaner.rb

```diff
@@ -197,6 +221,30 @@ def service_from_stats_key(stats_key)
             # simply ignore those keys.
             nil
           end
+
+          def delete_stats_keys_with_val_0(redis_conn, log_deleted_keys)
+            cursor = 0
+
+            loop do
+              cursor, keys = redis_conn.scan(cursor, count: SCAN_SLICE)
+
+              stats_keys = keys.select { |k| is_stats_key?(k) }
+
+              unless stats_keys.empty?
+                values = redis_conn.mget(*(stats_keys.to_a))
+                to_delete = Hash[stats_keys.zip(values)].select { |_k, v| v == '0'.freeze }
+
+                unless to_delete.empty?
+                  redis_conn.del(to_delete.keys)
```

Comment:
What if the `scan`, the `mget` or the `del` fails? You would break and not recover the original scan data from the point it broke. Given how much time these can take, would you want to try to recover a couple of times before giving up?

Reply:
The existing `delete!` method didn't have retries. I added the for both: https://github.com/3scale/apisonator/pull/250/commits/5d8a9e1a0d7ded5cf9c1d2bc028cf75b7251aa75